server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java

```diff
@@ -0,0 +1,155 @@
+        this.pattern = typeOf(JoinPlan.class)
+            .with(join ->
+                SUPPORTED_JOIN_TYPES.contains(join.joinType()) &&
+                !join.joinConditionsExtractedToFilters() &&
+                join.joinCondition() != null &&
+                hasConstantJoinConditions(join.joinCondition())
+            );
+    }
+
+    private static boolean hasConstantJoinConditions(Symbol joinCondition) {
+        for (var condition : QuerySplitter.split(joinCondition).values()) {
+            if (numberOfRelationsUsed(condition) <= 1) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Pattern<JoinPlan> pattern() {
+        return pattern;
+    }
+
+    @Override
+    public LogicalPlan apply(JoinPlan joinPlan,
+                             Captures captures,
+                             PlanStats planStats,
+                             TransactionContext txnCtx,
+                             NodeContext nodeCtx,
+                             UnaryOperator<LogicalPlan> resolvePlan) {
+        var conditions = joinPlan.joinCondition();
+        var allConditions = QuerySplitter.split(conditions);
+        var constantConditionsForFilter = new LinkedHashSet<Symbol>(allConditions.size());
+        var remainingConditions = new LinkedHashSet<Symbol>(allConditions.size());
+        var lhsRelations = new HashSet<>(joinPlan.lhs().relationNames());
+        var rhsRelations = new HashSet<>(joinPlan.rhs().relationNames());
+        for (var condition : allConditions.entrySet()) {
+            Set<RelationName> relationsInCurrentConditions = RelationNameCollector.collect(condition.getValue());
+            if (relationsInCurrentConditions.size() == 1) {
+                // Constant join conditions can only pushed to the preserved side of a join, where no null values are created:
+                // - lhs/rhs sides of cross/inner joins
+                // - lhs of left join
+                // - rhs of right join
+                if (joinPlan.joinType().isOuter()) {
+                    var matchesLhs = Sets.intersection(lhsRelations, relationsInCurrentConditions);
+                    var matchesRhs = Sets.intersection(rhsRelations, relationsInCurrentConditions);
+                    if ((!matchesLhs.isEmpty() && joinPlan.joinType() == LEFT) ||
+                        (!matchesRhs.isEmpty() && joinPlan.joinType() == RIGHT)) {
+                        constantConditionsForFilter.add(condition.getValue());
+                    } else {
+                        remainingConditions.add(condition.getValue());
+                    }
+                } else {
+                    constantConditionsForFilter.add(condition.getValue());
+                }
+            } else {
+                remainingConditions.add(condition.getValue());
+            }
+        }
+        if (!constantConditionsForFilter.isEmpty() && !remainingConditions.isEmpty()) {
```

Comment:
I think the `!remainingConditions.isEmpty()` should be removed, e.g:
```
SELECT * FROM t1 JOIN t2 ON t1.a > 10 AND t1.a > 1
```
here we should extract both as a filter which will be pushed down to t1 afterwards.
But maybe it needs further processing, as the JOIN currently will remain `INNER` where essentially it should become `CROSS`. I think the same goes for LEFT/RIGHT, they become `CROSS`.

It can definitely be addressed as follow-up, since it's an edge case where there are no conditions involving cols from the 2 tables.